### PR TITLE
feat(resources) add http-project-guide

### PR DESCRIPTION
**Pull request by Moe-Szyslak**
Big chunk of content.

All pulled in from https://github.com/jkbbwr/http-project-guide

This project will give you a a-z guide to help you build a http server

### CODE REVIEW

File: src/content/resources/topic/projects/http-project-guide/intro.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+Feel free to ask for help, search topics online, look for tutorials for any of the concepts presented.
+
+The idea is to guide you along but not give you all the answers. There will be pseudo code in some of the chapters in 
+this project its down to you to achieve the functionality in your language of choice. 
+The pseudo code isn't representative of a final project and is there to convey specific ideas you need to achieve
+
+## Rules
+
+- Put in the work yourself. Don't copy and paste, the only person you are cheating is yourself.
+- If you are helping don't spoon-feed. Its about the journey not the destination.
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/intro.md:30]: grammar: It's

File: src/content/resources/topic/projects/http-project-guide/chapter-06.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+---
+
+## Chapter 6: What is a protocol?
+How do we humans communicate? Through a mutually agreed convention of a language and conventions.
+Similarly, How do we make clients and servers talk to each other? 
+They need some sort of agreed way to ask for things and respond with replies or errors. This mutually agreed set of rules is commonly known as a protocol.
+
+In this chapter, we are going to create a really simple protocol for getting and storing strings, a *dictionary*.
+With our hypothetical protocol your server will store the definition to a series of words.
+You can do this nice and simply with a map like data structure or something more complex, your choice.
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-06.md:15]: map-like maybe?

File: src/content/resources/topic/projects/http-project-guide/chapter-06.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+```
+or 
+```txt
+ERROR can't find someword
+```
+We have created a simple reading protocol. This protocol can be extended by adding new requests of the form:
+```txt
+VERB args go here
+```
+I encourage you to mess about and add other verb based commands.
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-06.md:40]: mess around?

File: src/content/resources/topic/projects/http-project-guide/chapter-05.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
+---
+authors:
+  - "Kibb#4205"
+created_at: 2020/06/08
+title: "Chapter 4: Reading from the socket"
+---
+
+## Chapter 4: Reading from the socket.
+
+Its a good start, we have a server that can accept connections. It will write a message to the client.
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-05.md:10]: grammar: it's

File: src/content/resources/topic/projects/http-project-guide/chapter-02.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+to be building a server!
+
+What is a server? Its a program that accepts incoming connections, reads and writes to that connection and closing 
+it eventually.
+
+You will hear the term socket used a lot in this project. Sockets are a way of connecting two devices on a network together.
+One socket listens, the other socket connects.
+
+The main design of sockets in the unix world is [Berkeley Sockets](https://en.wikipedia.org/wiki/Berkeley_sockets) and
+on windows its [Winsock](https://www.wikiwand.com/en/Winsock).
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-02.md:20]: grammar: it's

File: src/content/resources/topic/projects/http-project-guide/chapter-02.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+created_at: 2020/06/08
+title: "Chapter 2: So what are we building?"
+---
+
+## Chapter 2: So what are we building?
+
+The actual final end goal is going to remain a mystery for the time being, but what I can tell you is that we are going
+to be building a server!
+
+What is a server? Its a program that accepts incoming connections, reads and writes to that connection and closing
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-02.md:13]: grammar: it's

File: src/content/resources/topic/projects/http-project-guide/chapter-10.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+`OPTIONS` is kind of a query to see what is allowed on a given resource, it should reply with what methods are supported and any other information.
+
+`TRACE` and `CONNECT` we are going to ignore these two as they are not commonly used anymore and wont be of much use to us.
+
+
+### Resource
+
+A resource is just something identified by a URL. An example of a resource could be `/index.html` or `/api/person`
+These are both resources, its down to the server to decide what these resources mean. In a web application they might
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-10.md:47]: grammar: it's

File: src/content/resources/topic/projects/http-project-guide/chapter-12.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+nulla diam imperdiet enim, id pellentesque leo nibh et urna. Vivamus non tortor dapibus, efficitur ex sed, 
+blandit odio. Vivamus volutpat, nunc sed pulvinar pellentesque, ipsum ante vestibulum sem, vitae malesuada 
+dui odio ut erat.
+```
+
+Boom. Done.
+
+
+There are other replies we might send though, for example lets say someone requests `missing.txt` and we don't have that.
+We need to inform the client it doesn't exist. We do this by sending a 404. Its mostly the same as the previous but we change
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-12.md:69]: grammar: it's

File: src/content/resources/topic/projects/http-project-guide/chapter-13.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+authors:
+  - "Kibb#4205"
+created_at: 2020/06/08
+title: "Chapter 12: Config"
+---
+
+## Chapter 12: Config.
+
+For the time being we have been hard coding all the various values like the port and the web directory, 
+its about time we made this configurable.
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-13.md:11]: grammar: it's

File: src/content/resources/topic/projects/http-project-guide/chapter-13.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+web_directory: ./www
+```
+
+Lets update the program to read the values from config and use those.
+
+The whole point of making it configurable is the person deploying this in the future doesn't want to edit code or 
+change values in your program to get it to work to their usecase. So by providing a way to configure your application 
+it makes it more portable and usable in different situations.
+
+This should be quite an easy chapter but its a reasonably important one.
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-13.md:30]: grammar: it's

File: src/content/resources/topic/projects/http-project-guide/intro.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+You can use any language that the server currently has a channel for.
+
+This project will expect you to have basic understanding of the language you are using and have a few extra tools 
+installed like; `git`, `netcat` (linux), `putty` (windows)
+
+This project will guide you in the direction of what you need to do but it leaves individual learning up to you.
+Feel free to ask for help, search topics online, look for tutorials for any of the concepts presented.
+
+The idea is to guide you along but not give you all the answers. There will be pseudo code in some of the chapters in 
+this project its down to you to achieve the functionality in your language of choice.
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/intro.md:24]: grammar: it's

File: src/content/resources/topic/projects/http-project-guide/chapter-01.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+You should pick a suitable name for this project, you 
+can pick a word or phrase yourself or use a [project name generator](http://codenames.herokuapp.com/).
+Try to make sure that there is no space in your project name as this will cause issues later.
+
+If you are using a build tool that sets up the environment for you now would be a good time to run it.
+
+Feel free to add the Hello World of your language to test that everything is setup correctly.
+
+One of the key steps in this project will be to track progress and changes as you develop. 
+For this we are going to be using `git` if you do not know git,
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-01.md:22]: comma or period after `git`

File: src/content/resources/topic/projects/http-project-guide/chapter-07.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+```js
+fun test_get_definition() {
+    var client = connect("127.0.0.1", 5678);
+    client.send("GET word\n");
+    var line = client.recv(4096);
+    assert(line == "ANSWER something interesting here\n");
+}
+```
+The idea is to try and test all the functionality of your program and use these tests to identify what is going wrong.
```
Comment (veksen) [inline src/content/resources/topic/projects/http-project-guide/chapter-07.md:26]: those or these?

Comment (Moe-Szyslak) [reply src/content/resources/topic/projects/http-project-guide/chapter-07.md:26]: I think these is correct here, its speaking in present tense

Comment (Moe-Szyslak) [reply src/content/resources/topic/projects/http-project-guide/chapter-07.md]: For sake of clarity, reworded that sentence to:
```The idea here is to try and test all the functionality of your program and use the above tests to identify what may be going wrong.```

**Pull request merged**